python_modules/dagster/dagster_tests/core_tests/definitions_tests/decorators_tests/test_op.py

```diff
@@ -487,3 +493,113 @@ def nested_ops():
         "{'config': {'foo': '...'}}}}}",
     ):
         nested_job.execute_in_process()
+
+
+def test_log_events():
+    @op
+    def basic_op(context):
+        context.log_event(AssetMaterialization("first"))
+        context.log_event(Materialization("second"))
+        context.log_event(AssetMaterialization("third"))
+        context.log_event(ExpectationResult(success=True))
+        context.log_event(AssetObservation("fourth"))
+
+    with instance_for_test() as instance:
+        result = execute_op_in_graph(basic_op, instance=instance)
+        asset_materialization_keys = [
+            materialization.label
+            for materialization in result.asset_materializations_for_node("basic_op")
+        ]
+
+        assert asset_materialization_keys == ["first", "second", "third"]
+
+        relevant_events_from_execution = [
+            event
+            for event in result.all_node_events
+            if event.is_asset_observation
+            or event.is_step_materialization
+            or event.is_expectation_result
+        ]
+
+        relevant_events_from_event_log = [
+            event_log.dagster_event
+            for event_log in instance.all_logs(result.run_id)
+            if event_log.dagster_event is not None
+            and (
+                event_log.dagster_event.is_asset_observation
+                or event_log.dagster_event.is_step_materialization
+                or event_log.dagster_event.is_expectation_result
+            )
+        ]
+
+        def _assertions_from_event_list(events):
+            assert events[0].is_step_materialization
+            assert events[0].event_specific_data.materialization.label == "first"
+
+            assert events[1].is_step_materialization
+            assert events[1].event_specific_data.materialization.label == "second"
+
+            assert events[2].is_step_materialization
+            assert events[2].event_specific_data.materialization.label == "third"
+
+            assert events[3].is_expectation_result
+
+            assert events[4].is_asset_observation
+
+        _assertions_from_event_list(relevant_events_from_execution)
+
+        _assertions_from_event_list(relevant_events_from_event_log)
+
+
+def test_yield_event_ordering():
+    @op
+    def yielding_op(context):
+        context.log_event(AssetMaterialization("first"))
+        time.sleep(1)
+        context.log.debug("A log")
+        context.log_event(AssetMaterialization("second"))
+        yield AssetMaterialization("third")
+        yield Output("foo")
```

Comment:
there should be outro `context.log_event` to protect what [2] handles from accidental change